### PR TITLE
pre-commit: add black, isort, pyupgrade, autoflake, flake8, and misc hooks

### PR DESCRIPTION
**Pull request by consideRatio**
This adds the usual hooks except prettier to this repo. I'm omitting prettier for now since I expect it could lead to merge conflicts with docs PRs #99.

This PR will have a probably quite simple to resolve merge conflict with #90.

### CODE REVIEW

**Review by yuvipanda — CHANGES_REQUESTED** [review]
I'm kinda wary of ruff given it's VC funding, and would prefer we use the same stack we use elsewhere instead (black, etc). Nothing against the people behind ruff specifically! If we want to introduce ruff into the jupyterhub project, I'd like there to be more community consensu before we do so.

**Comment by consideRatio** [comment]
Ah okay, we can go for the normal stuff - I adopted it after setting it be piloted for use in jupyterhub/jupyterhub

File: dashboards/jupyterhub.jsonnet [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env jsonnet -J ../vendor
+#!/usr/bin/env -S jsonnet -J ../vendor
```
Comment (consideRatio) [inline dashboards/jupyterhub.jsonnet:1]: I had to use `-S` to not have env parse the content above as a command with spaces in it.

File: dashboards/support.jsonnet [hunk 2]
```diff
@@ -1,3 +1,4 @@
+#!/usr/bin/env -S jsonnet -J ../vendor
```
Comment (consideRatio) [inline dashboards/support.jsonnet:1]: I had to ensure shebangs were added to executable files.

I removed the executable flag for the libsonnet file that renders blank anyhow.

**Comment by consideRatio** [comment]
@yuvipanda I'll go for a merge now to unblock myself to contribute further, I'll push a merge commit to your PR fixing the conflict caused by this as well

**Pull request merged**